rinex-cli/src/cli/mod.rs

```diff
@@ -0,0 +1,343 @@
+use clap::{value_parser, Arg, ArgAction, ArgMatches, ColorChoice, Command};
+use log::info;
+use map_3d::{ecef2geodetic, geodetic2ecef, Ellipsoid};
+use std::path::{Path, PathBuf};
+use std::str::FromStr;
+
+use crate::Error;
+use rinex::prelude::*;
+use walkdir::WalkDir;
+
+// identification mode
+mod identify;
+// graph mode
+mod graph;
+// merge mode
+mod merge;
+// split mode
+mod split;
+// tbin mode
+mod time_binning;
+// QC mode
+mod qc;
+// positioning mode
+mod positioning;
+
+pub struct Cli {
+    /// Arguments passed by user
+    pub matches: ArgMatches,
+}
+
+impl Default for Cli {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// Context defined by User.
+pub struct Context {
+    /// Data context defined by user
+    pub data: RnxContext,
+    /// Quiet option
+    pub quiet: bool,
+    /// Workspace is the place where this session will generate data.
+    /// By default it is set to $WORKSPACE/$PRIMARYFILE.
+    /// $WORKSPACE is either manually definedd by CLI or we create it (as is).
+    /// $PRIMARYFILE is determined from the most major file contained in the dataset.
+    pub workspace: PathBuf,
+    /// (RX) reference position to be used in further analysis.
+    /// It is either (priority order is important)
+    ///  1. manually defined by CLI
+    ///  2. determined from dataset
+    pub rx_ecef: Option<(f64, f64, f64)>,
+}
+
+impl Context {
+    /*
+     * Utility to determine the most major filename stem,
+     * to be used as the session workspace
+     */
+    pub fn context_stem(data: &RnxContext) -> String {
+        let ctx_major_stem: &str = data
+            .rinex_path()
+            .expect("failed to determine a context name")
+            .file_stem()
+            .expect("failed to determine a context name")
+            .to_str()
+            .expect("failed to determine a context name");
+        /*
+         * In case $FILENAME.RNX.gz gz compressed, we extract "$FILENAME".
+         * Can use .file_name() once https://github.com/rust-lang/rust/issues/86319  is stabilized
+         */
+        let primary_stem: Vec<&str> = ctx_major_stem.split('.').collect();
+        primary_stem[0].to_string()
+    }
+    /*
+     * Creates File/Data context defined by user.
+     * Regroups all provided files/folders,
+     */
+    pub fn from_cli(cli: &Cli) -> Result<Self, Error> {
+        let mut data = RnxContext::default();
+        /* load all directories recursively, one by one */
+        for dir in cli.input_directories() {
+            let walkdir = WalkDir::new(dir).max_depth(5);
```

Comment:
Should probably be configurable in the long run.

Reply:
It does not cost much to add this into the general panel.
Something like `--depth X` to make the `-d` flag more flexible.
I will add it